src/test/app/TxQ_test.cpp

```diff
@@ -610,24 +869,15 @@ class TxQ_test : public beast::unit_test::suite
         }
         checkMetrics(env, 4, 8, 5, 4, 256);
 
-        // This transaction also has an "infinite" fee level,
-        // but since bob has txns in the queue, it gets queued.
+        // Since bob has entries in his queue he cannot add a blocker.
         env(regkey(bob, alice), fee(0),
-            seq(seqBob), queued);
-        ++seqBob;
-        checkMetrics(env, 5, 8, 5, 4, 256);
-
-        // Unfortunately bob can't get any more txns into
-        // the queue, because of the multiTxnPercent.
-        // TANSTAAFL
-        env(noop(bob), fee(XRP(100)),
-            seq(seqBob), ter(telINSUF_FEE_P));
-
-        // Carol fills the queue, but can't kick out any
-        // transactions.
+            seq(seqBob), ter(telCAN_NOT_QUEUE_BLOCKS));
```

Comment:
So I'm digging into this test now.  I don't quite have the changes you want yet.  Here's what I did:

1. Rather than filling bob's queue prior to putting in the `regkey`, I'm filling Alice's queue.
2. Then I submit bob's `regkey`, expecting it to go into the queue.  Instead it goes straight into the ledger.  The new `tryDirectSubmit` sends the `regkey` straight in because it has a very high (but not infinite) `feeLevelPaid`.
3. I do more stuff with carol, but that's irrelevant since the `regkey` is not in the queue.

Since `regkey` is a blocker and must be alone in bob's queue I'm unsure how to keep bob's `regkey` in the queue.  Maybe this indicates a problem with the `tryDirectSubmit` implementation?

Thanks for your thoughts.

Reply:
@scottschurr  We've discussed this offline. Here's a very short summary of the conversation.

Basically, there are feasible ways to solve this problem, specifically in the test.
1. Override the value of `zero_basefee_transaction_feelevel` for this test to be something more in the range of Alice's queued transactions, causing the tx to go into the queue.
2. Refactor the way these zero-fee transactions are handled internally so that the system never has to deal with the divide-by-zero issue. One way to implement this would be for `CalculateBaseFee` to return the actual base fee, and an "effective fee paid". In most cases, the base fee would remain unchanged, and the "effective fee" would be the actual fee. In this special case, setting a regkey with no fee required, the base fee would be the default fee as if this was a normal transaction, but the effective fee would be `the actual fee + toDrops(base fee)`. Any place that uses the base fee and the transaction fee would instead use the effective fee. Everywhere else, particularly when the fee is actually charged, would use the actual fee.